### DISCUSSION TITLE
warp_by_scalar error.

### DISCUSSION THREAD

**Discussion by Davegis16** [pyvista/pyvista#2248] [Q&A]
**Qu 1.** Has anyone had problems with reading in Surfer grid file and applying a warp_scaler.

**Using the following results in an error below.**
Python 3.89
pyvista 0.33.2
pvgeo 2.1.0

_Traceback (most recent call last):_

  _File "E:\Program_Files\WinPy64-3890\Projects\OMF\Surfer grids\Convert grd to VTU add to PVD .py", line 36, in <module>
    surfergrid = dem.warp_by_scalar(scale_factor=1.0)_

  _File "E:\Program_Files\WinPy64-3890\python-3.8.9.amd64\lib\site-packages\pyvista\core\filters\data_set.py", line 1981, in warp_by_scalar_
    _arr = get_array(self, scalars, preference='point', err=True)_

  _File "E:\Program_Files\WinPy64-3890\python-3.8.9.amd64\lib\site-packages\pyvista\utilities\helpers.py", line 374, in get_array
    raise KeyError(f'Data array ({name}) not present in this dataset.')_

_KeyError: 'Data array (None) not present in this dataset.'_

**Using the same code and the following, warp_scaler works fine.**
Python 3.76
pyvista 0.31.3
pvgeo 2.1.0 

**Qu 2.** Is there any documentation or examples of Surfer or ESRi grids that are imported with cells with no data, which would normally 
         mark the cell to be rendered invisible and how it can be handle in Pyvista.
         
For example the white cells around this surface
![image](https://user-images.githubusercontent.com/73296209/154863836-39f538f2-1728-42b3-8ca3-a6594e26a0e9.png)

Thanks

Dave

**Comment by RichardScottOZ** [comment]
Do you gave a few? Same behaviour on each? Never tried one, have to see if one exists in our archive.

**Comment by trhoye** [comment]
python 3.10.4
pyvista  0.33.3
pvgeo  1.1.41
vtk       9.1.0

I just searched for this error.
Yes I've got the same problem - a program I wrote a year ago reading surfer grids.
Do not have the answer yet! Apparently the error is linked to the vtk module.

Error message
AttributeError: 'vtkmodules.vtkCommonDataModel.vtkImageData' object has no attribute 'warp_by_scalar'

I got a similar error on:
us1 = SurferGridReader().apply(fname1)
AttributeError: 'SurferGridReader' object has no attribute 'apply'
This error message disappeared by using a capital A in Apply (?)

Did you find the solution re. warp-by-scalar?

**Reply by Davegis16** [reply]
I gave up on finding a solution, one problem is Surfer grids files have a header file which shows the maximum and minimum elevation range for the grid. When you have a empty cells (blanking), the value used to represent the blanked cell falls outside of this range and the warp by vector does not like it. You can manual change the minimum value in the header of ascii surfer files, then filter using the threshold.

In the end I saved grids from Surfer as Geotiffs and used this code to open them (and used threshold to remove empty cells)
https://github.com/pyvista/pyvista-support/issues/205

**Reply by RichardScottOZ** [reply]
Right, so data format issues

**Reply by banesullivan** [reply]
Make sure to upgrade to PVGeo 2.x
